server/controllers/users.js

```diff
@@ -92,25 +96,50 @@ export const updateToken = async (req, res) => {
 };
 
 /**
- * Verify the 2FA code the user has entered when
- * logging in and send back another JWT.
+ * Verify the 2FA code or recovery code the user has entered when logging in and send back another JWT.
  */
 export const twoFactorAuthAndUpdateToken = async (req, res, next) => {
-  const { twoFactorAuthenticatorCode } = req.body;
+  const { twoFactorAuthenticatorCode, twoFactorAuthenticationRecoveryCode } = req.body;
 
   const userId = Number(req.jwtPayload.sub);
+
+  // Both 2FA and recovery codes rate limited to 10 tries per hour
+  const rateLimit = new RateLimit(`user_2FA_endpoint_${userId}`, 10, ONE_HOUR_IN_SECONDS);
+  if (!(await rateLimit.registerCall())) {
+    return next(new Unauthorized('Too many attempts. Please try again in an hour'));
+  }
+
```

Comment:
:ice_cream: Since admins can be forced to update their tokens multiple time when they pay expenses, doing that here might be a too strong limitation. I would only count in the rate limit. One simple way to achieve that could be to have something like this:

```suggestion
  // Both 2FA and recovery codes rate limited to 10 tries per hour
  const rateLimit = new RateLimit(`user_2FA_endpoint_${userId}`, 10, ONE_HOUR_IN_SECONDS);
  const fail = (exception) => {
    await rateLimit.registerCall();
    next(exception);
  }

  if (await rateLimit.hasReachedLimit()) {
    return next(new Unauthorized('Too many attempts. Please try again in an hour'));
  }
```

Then replace all the `return next(new Exception(...))` by `return fail(new Exception(...))`. For example:
```es6
return fail(new Unauthorized('Two-factor authentication code failed. Please try again'));
```